test/framework/klusterlet.go

```diff
@@ -0,0 +1,259 @@
+		// create external-managed-kubeconfig, will use the same cluster to simulate the Hosted mode.
+		secret.Namespace = agentNamespace
+		secret.Name = helpers.ExternalManagedKubeConfig
+		if _, err := spoke.KubeClient.CoreV1().Secrets(agentNamespace).Get(context.TODO(), secret.Name, metav1.GetOptions{}); err != nil {
+			if !apierrors.IsNotFound(err) {
+				klog.Errorf("failed to get secret %v in ns %v. %v", secret.Name, agentNamespace, err)
+				return nil, err
+			}
+			if _, err = spoke.KubeClient.CoreV1().Secrets(agentNamespace).Create(context.TODO(), secret, metav1.CreateOptions{}); err != nil {
+				klog.Errorf("failed to create secret %v in ns %v. %v", secret, agentNamespace, err)
+				return nil, err
+			}
+		}
+	}
+
+	// create klusterlet CR
+	realKlusterlet, err := spoke.OperatorClient.OperatorV1().Klusterlets().Create(context.TODO(),
+		klusterlet, metav1.CreateOptions{})
+	if err != nil && !apierrors.IsAlreadyExists(err) {
+		klog.Errorf("failed to create klusterlet %v . %v", klusterlet.Name, err)
+		return nil, err
+	}
+
+	return realKlusterlet, nil
+}
+
+func CreatePureHostedKlusterlet(spoke *Spoke, name, clusterName string) (*operatorapiv1.Klusterlet, error) {
+	if name == "" {
+		return nil, fmt.Errorf("the name should not be null")
+	}
+
+	var klusterlet = &operatorapiv1.Klusterlet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: operatorapiv1.KlusterletSpec{
+			RegistrationImagePullSpec: "quay.io/open-cluster-management/registration:latest",
+			WorkImagePullSpec:         "quay.io/open-cluster-management/work:latest",
+			ExternalServerURLs: []operatorapiv1.ServerURL{
+				{
+					URL: "https://localhost",
+				},
+			},
+			ClusterName: clusterName,
+			DeployOption: operatorapiv1.KlusterletDeployOption{
+				Mode: operatorapiv1.InstallModeHosted,
+			},
+		},
+	}
+
+	// create klusterlet CR
+	realKlusterlet, err := spoke.OperatorClient.OperatorV1().Klusterlets().Create(context.TODO(),
+		klusterlet, metav1.CreateOptions{})
+	if err != nil {
+		klog.Errorf("failed to create klusterlet %v . %v", klusterlet.Name, err)
+		return nil, err
+	}
+
+	return realKlusterlet, nil
+}
+
+func CheckKlusterletStatus(spoke *Spoke, klusterletName, condType, reason string, status metav1.ConditionStatus) error {
+	klusterlet, err := spoke.OperatorClient.OperatorV1().Klusterlets().Get(context.TODO(), klusterletName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	cond := meta.FindStatusCondition(klusterlet.Status.Conditions, condType)
+	if cond == nil {
+		return fmt.Errorf("cannot find condition type %s", condType)
+	}
+
+	if cond.Reason != reason {
+		return fmt.Errorf("condition reason is not matched, expect %s, got %s", reason, cond.Reason)
+	}
+
+	if cond.Status != status {
+		return fmt.Errorf("condition status is not matched, expect %s, got %s", status, cond.Status)
+	}
+
+	return nil
+}
+
+// CleanKlusterletRelatedResources needs both hub side and spoke side operations
+func CleanKlusterletRelatedResources(
```

Comment:
we actually should have a clean func for spoke and hub. Something like spoke.Cleanup() and hub.Cleaup, and call these two in this func.

Reply:
Agree, this is something simply copied and pasted in the PR.

But in this case, it's not like hub.Cleanup is suitable, because hub.Cleanup is like a entire hub Cleanup, in this function, we only want to clean up the managedcluster created by the klusterlet on the hub.

And also `spoke != klustelret`, spoke has multiple klusterlet during the e2e.

I think we can have hub.Cleanup and spoke.Cleanup but not for this case.

Reply:
this could be confusing, because to me a spoke == a klusterlet

Reply:
and for hub, hub.Cleanup(cluster1) should be fine.

Reply:
During the e2e we are deploying multiple klusterlet on the same spoke: 

https://github.com/open-cluster-management-io/ocm/blob/7a8e444a52a360a6425bb9a2b01746e0e0c2f7c9/test/e2e/klusterlet_test.go#L38

In our e2e, a spoke should be a cluster that has klusterlet operator on it, and we deploy multiple klusterlet on one spoke.

Reply:
Should we use `CleanUpManagedCluster`? Because if we use `CleanUp` specifically represent to clean up a managedcluster, then clean up manifestwork, clustermanager and other resources will be confusing.

Reply:
I think you will need:
1: just clean klusterlet CR,
2. clean klusterlet CR and klusterlet operator
3. just clean a managedcluster CR
4. clean them all

Reply:
Yes, can I list it as a TODO in the future, because "clean the klusterlet operator" part feels need more discussion, it's a behavior change comparing with previous code.